server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java

```diff
@@ -0,0 +1,203 @@
+            return shardId == that.shardId
+                && snapshot.equals(that.snapshot)
+                && index.equals(that.index);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(snapshot, index, shardId);
+        }
+
+        @Override
+        public String toString() {
+            return "SnapshotShard{" +
+                "snapshot=" + snapshot +
+                ", index=" + index +
+                ", shardId=" + shardId +
+                '}';
+        }
+    }
+
+    @Override
+    public SnapshotShardSizeInfo snapshotShardSizes() {
+        return new SnapshotShardSizeInfo(snapshotShardSizes);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster()) {
+            final Set<SnapshotShard> requiredSnapshotShards = new HashSet<>();
+            for (ShardRouting shardRouting : event.state().routingTable().shardsWithState(ShardRoutingState.UNASSIGNED)) {
+                if (shardRouting.primary() && shardRouting.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
+                    final RecoverySource.SnapshotRecoverySource snapshotRecoverySource =
+                        (RecoverySource.SnapshotRecoverySource) shardRouting.recoverySource();
+                    final SnapshotShard snapshotShard = new SnapshotShard(snapshotRecoverySource.snapshot(),
+                        snapshotRecoverySource.index(), shardRouting.id());
+                    requiredSnapshotShards.add(snapshotShard);
+                    // check if already populated entry
+                    if (snapshotShardSizes.get(snapshotShard) == null) {
+                        // check if already fetching snapshot info in progress
+                        if (fetchingShards.add(snapshotShard)) {
+                            // TODO: Use a dedicated threadpool here? Use FETCH thread pool?
+                            threadPool.generic().execute(new AbstractRunnable() {
+                                @Override
+                                public void onFailure(Exception e) {
+                                    logger.warn(new ParameterizedMessage("failed to retrieve shard size information for {}",
+                                        shardRouting), e);
+                                }
+
+                                @Override
+                                protected void doRun() {
+                                    final RepositoriesService repositories = repositoriesServiceSupplier.get();
+                                    assert repositories != null;
+                                    final Repository repository = repositories.repository(snapshotShard.snapshot.getRepository());
+                                    final IndexShardSnapshotStatus status =
+                                        repository.getShardSnapshotStatus(snapshotRecoverySource.snapshot().getSnapshotId(),
```

Comment:
I think this will lead to a burst of many requests to the blob store to get the shard size whenever a node holding searchable snapshots fails or is restarted.

This could block the generic thread pool for a while, potentially causing other issues. Similarly if moving to the FETCH_SHARD_STORE pool it might interfere with recoveries of primaries.

Generic pool has minimum 128 threads, which does mean a decent amount of parallel fetching from blobstore, but going for other pools would likely lower this considerably.

It also feels like a bad spot in our allocation to have a dependency on an external blob store, in particular if it responds very slowly for odd reasons. Once persistent cache is in place, we should be able to recover on our own without accessing the blob store.

Reply:
++ we must limit the concurrency here. This isn't just about threads (though that's reason enough) but also about memory to some degree.  `org.elasticsearch.repositories.blobstore.BlobStoreRepository#getShardSnapshotStatus` might use a few MB per shard in memory because the underlying metadata file that is loaded from the blobstore is non-trivial in size (we essentially have the same requests in the `_status` APIs for snapshot status and don't parallelize them there either at least for now because we have to be careful about the concurrency).